api/controllers/game/rematch.js

```diff
@@ -40,14 +40,37 @@ module.exports = async function (req, res) {
       shouldNewGameBeRanked,
       gameService.GameStatus.STARTED,
     );
-    await Promise.all([
+    const [ , , p0, p1] = await Promise.all([
       Game.updateOne({ id: updatedGame.id }).set({ rematchGame: newGame.id }),
       Game.replaceCollection(newGame.id, 'players').members([newP0Id, newP1Id]),
       User.updateOne({ id: newP0Id }).set({ pNum: 0 }),
       User.updateOne({ id: newP1Id }).set({ pNum: 1 }),
     ]);
     updatedGame.rematchGame = newGame.id;
 
+    //Get all exisiting rematchGames
+    let rematchGames = [];
+    const getRematchGames = async (gameId) => {
+      const [gameToAdd] = await Game.find({ rematchGame: gameId });
+      if (gameToAdd) {
+        rematchGames.unshift(gameToAdd);
```

Comment:
My initial reaction was that it would be tidier if this function returned the array instead of mutating it as a side effect like
```
const getRematchGames = async (gameId, previousGames) () => {
  const gameToAdd = await Game.findOne({ rematchGame: gameId });
  if (gameToAdd) {
    previousGames.unshift(gameToAdd);
    if (gameToAdd.rematchGame) {
      previousGames = await getRematchGames(gameToAdd.id, previousGames);
    }
  }
  return previousGames;
}
```

But now I wonder about the performance implications of each recursive function having its own separate reference to the array being mutated. I suspect your implementation is more performant, and it still reads reasonably clearly. 

@seriouslysean what do you think?

Reply:
Honestly, I'm a fan of less loops and this is a clean implementation of a recursive function. If I were writing it, I'd prefer fast fails to what we have here, but I wouldn't really change much else -- this is really nice!

```
// Get all existing rematchGames
const rematchGames = [];
const getRematchGames = async (gameId) => {
  const gameToAdd = await Game.findOne({ rematchGame: gameId });
  if (!gameToAdd) {
    return;
  }
  rematchGames.unshift(gameToAdd);
  if (!gameToAdd.rematchGame) {
    return;
  }
  await getRematchGames(gameToAdd.id);
};
```

I would just be very clear about any pitfalls with this approach as I don't have the full context. Make sure it's not able to run forever, test it well and look for any other areas you may need to exit cleanly. Do you need any sort of tray watch if what await fails inside of the rematchGame check, etc.

Reply:
I started with a fast fail but it didn't look as clean visually imo

Reply:
I’d say switch it, it’s logically more coherent and generally more performant.

Reply:
Cool, okay will do this when I get home from work later 🙂